app/models/ArrivalUpdate.scala

```diff
@@ -17,11 +17,44 @@
 package models
 
 import cats._
-import play.api.libs.json.JsObject
 import play.api.libs.json.Json
 import play.api.libs.json.Writes
 
-case class MessageStatusUpdate(messageId: MessageId, messageStatus: MessageStatus)
+sealed trait ArrivalUpdate
+
+object ArrivalUpdate {
+
+  implicit val arrivalUpdateSemigroup: Semigroup[ArrivalUpdate] = {
+    case (ArrivalStatusUpdate(_), x @ ArrivalStatusUpdate(_))        => x
+    case (a @ ArrivalStatusUpdate(_), m @ MessageStatusUpdate(_, _)) => CompoundStatusUpdate(a, m)
+    case (ArrivalStatusUpdate(_), c @ CompoundStatusUpdate(_, _))    => c
+    case (ArrivalStatusUpdate(_), p @ ArrivalPutUpdate(_, _))        => p
+
+    case (MessageStatusUpdate(_, _), x @ MessageStatusUpdate(_, _))  => x
+    case (m @ MessageStatusUpdate(_, _), a @ ArrivalStatusUpdate(_)) => CompoundStatusUpdate(a, m)
+    case (MessageStatusUpdate(_, _), c @ CompoundStatusUpdate(_, _)) => c
+    case (MessageStatusUpdate(_, _), p @ ArrivalPutUpdate(_, _))     => p
+
+    case (CompoundStatusUpdate(_, _), x @ CompoundStatusUpdate(_, _))    => x
+    case (c @ CompoundStatusUpdate(_, _), a @ ArrivalStatusUpdate(_))    => c.copy(arrivalStatusUpdate = a)
+    case (c @ CompoundStatusUpdate(_, _), m @ MessageStatusUpdate(_, _)) => c.copy(messageStatusUpdate = m)
+    case (CompoundStatusUpdate(_, _), p @ ArrivalPutUpdate(_, _))        => p
+
+    case (ArrivalPutUpdate(_, _), x @ ArrivalPutUpdate(_, _))         => x
+    case (p @ ArrivalPutUpdate(_, _), a @ ArrivalStatusUpdate(_))     => p.copy(arrivalUpdate = p.arrivalUpdate.copy(arrivalStatusUpdate = a))
+    case (p @ ArrivalPutUpdate(_, _), m @ MessageStatusUpdate(_, _))  => p.copy(arrivalUpdate = p.arrivalUpdate.copy(messageStatusUpdate = m))
+    case (p @ ArrivalPutUpdate(_, _), c @ CompoundStatusUpdate(_, _)) => p.copy(arrivalUpdate = c)
+  }
+
+  implicit val arrivalUpdateArrivalModifier: ArrivalModifier[ArrivalUpdate] = {
+    case x @ MessageStatusUpdate(_, _)  => ArrivalModifier.toJson(x)
+    case x @ ArrivalStatusUpdate(_)     => ArrivalModifier.toJson(x)
+    case x @ CompoundStatusUpdate(_, _) => ArrivalModifier.toJson(x)
+    case x @ ArrivalPutUpdate(_, _)     => ArrivalModifier.toJson(x)
```

Comment:
Is this matcher needed if the outcome is always the same?

Reply:
This is allow implicit resolution to pick the correct value for the ArrivalModifier.toJson